tests/flow/test_intern_string.py

```diff
@@ -0,0 +1,429 @@
+
+class testInternStringReplication():
+    def __init__(self):
+        # skip test if we're running under Valgrind
+        if VALGRIND or SANITIZER:
+            Environment.skip(None) # valgrind is not working correctly with replication
+
+        self.env, self.db = Env(env='oss', useSlaves=True)
+        self.conn = self.env.getConnection()
+        self.graph = self.db.select_graph(GRAPH_ID)
+
+        self.source_con = self.env.getConnection()
+        self.replica_con = self.env.getSlaveConnection()
+
+        # force effects replication
+        self.db.config_set('EFFECTS_THRESHOLD', 0)
+
+        # Synchronous deletion
+        self.source_con.execute_command("GRAPH.CONFIG", "SET", 'ASYNC_DELETE', 'no')
+        self.replica_con.execute_command("GRAPH.CONFIG", "SET", 'ASYNC_DELETE', 'no')
+
+        # clear DB
+        self.conn.flushall()
+        self.source_con.execute_command("WAIT", "1", "0")
+
+    def tearDown(self):
+        # clear DB
+        self.source_con.flushall()
+        self.source_con.execute_command("WAIT", "1", "0")
+
+        assertStringPoolStats(self.source_con, 0, 0)
+        assertStringPoolStats(self.replica_con, 0, 0)
+
+    def query_and_wait(self, q, p=None):
+        if p is None:
+            p = {}
+
+        res = self.graph.query(q, p)
+
+        # the WAIT command forces master slave sync to complete
+        self.source_con.execute_command("WAIT", "1", "0")
+
+        return res
+
+    def test_intern_string_replication(self):
+        # both master and replica should be empty
+        assertStringPoolStats(self.source_con, 0, 0)
+        assertStringPoolStats(self.replica_con, 0, 0)
+
+        # replicate a node creation containing an intern string
+
+        s = LARGE_STRING
+
+        # create first node
+        p = {'s': s}
+        q = "CREATE ({value: intern($s)})"
+        self.query_and_wait(q, p)
+
+        assertStringPoolStats(self.source_con, 1, 1)
+        assertStringPoolStats(self.replica_con, 1, 1)
+
+        # replicate an addition of an intern string
+        q = "MATCH (n) SET n.s = intern($s)"
+        self.query_and_wait(q, p)
+
+        assertStringPoolStats(self.source_con, 1, 2)
+        assertStringPoolStats(self.replica_con, 1, 2)
+
+        # replicate deletion of an intern string
+        q = "MATCH (n) SET n.s = null"
+        self.query_and_wait(q)
+
+        assertStringPoolStats(self.source_con, 1, 1)
+        assertStringPoolStats(self.replica_con, 1, 1)
+
+        # replicate update of an intern string
+        q = "MATCH (n) SET n.value = intern('intern-string')"
+        self.query_and_wait(q)
+
+        assertStringPoolStats(self.source_con, 1, 1)
+        assertStringPoolStats(self.replica_con, 1, 1)
+
+        # replicate deletion if a node
+        q = "MATCH (n) DELETE n"
+        self.query_and_wait(q)
+
+        assertStringPoolStats(self.source_con, 0, 0)
+        assertStringPoolStats(self.replica_con, 0, 0)
+
```

Comment:
missing test for running multiple concurrent read queries that create intern strings and see there is no crashes or leak in the pool